### PR TITLE
Meta-class and fine-grained class in buffer for label drift detection

### PR DESCRIPTION
**Pull request by AlifAshrafee**
This is a rough implementation of storing both the metaclass and fine-grained class in the buffer. This is not the final implementation. I created this branch just for code review, we don't have to merge it. After finishing the whole implementation, I realized it could be done in a much simpler and cleaner way if we simply return the fine-grained targets as before and just return the meta-labels as an additional parameter. That way, we can keep the rest of the code unchanged, and simply handle the metaclasses for the label drifting datasets. I will work on this refinement after our meeting tomorrow (11:00 am GMT-4).

### CODE REVIEW

File: models/er.py [hunk 1]
```diff
@@ -43,7 +47,12 @@ def observe(self, inputs, labels, not_aug_inputs):
...
         self.opt.step()
 
-        self.buffer.add_data(examples=not_aug_inputs,
+        if original_targets is not None:
+            self.buffer.add_data(examples=not_aug_inputs,
+                             labels=labels[:real_batch_size],
+                             original_labels=original_targets[:real_batch_size])
+        else: 
+            self.buffer.add_data(examples=not_aug_inputs,
                              labels=labels[:real_batch_size])
```
Comment (jedrzejkozal) [inline models/er.py]: instead putting of whole code inside if you can do:
original_targets = original_targets[:real_batch_size] if original_targets is not None
self.buffer.add_data(examples=not_aug_inputs,
                             labels=labels[:real_batch_size],
                             original_labels=original_targets)

original_labels in add_data is by default None, so you will be fine with it

File: utils/drift_detection.py [hunk 2]
```diff
@@ -108,14 +108,13 @@ def detect_uncertainty_label_drift(dataset, train_loader, model):
 
         new_images = torch.cat(filtered_images, dim=0)
 
-    if new_images.size(0) > 0 and hasattr(model, 'buffer'):
-        for meta_cls in dataset.METACLASSES:
```
Comment (jedrzejkozal) [inline utils/drift_detection.py:112]: remove METACLASSES from seq_fashionmnist and cifar100-label drift

File: utils/drift_detection.py [hunk 3]
```diff
@@ -96,7 +96,7 @@ def detect_uncertainty_label_drift(dataset, train_loader, model):
     drifting_classes = dataset.drifting_classes
```
Comment (jedrzejkozal) [inline utils/drift_detection.py]: instead of having two very similar functions we can make label_drift as additional keyword argument to detect_uncertainty_drift

File: utils/buffer.py [hunk 4]
```diff
@@ -278,15 +283,19 @@ def get_class_sample_count(self, label: int) -> torch.Tensor:
...
         If label not present in the buffer, then raise ValueError exception.
         """
-        idx = torch.argwhere(self.labels != label).flatten()
-        num_removed = len(self.labels) - len(idx)
+        if metaclass:
+            idx = torch.argwhere(self.original_labels != label).flatten()
+            num_removed = len(self.original_labels) - len(idx)
+        else:
+            idx = torch.argwhere(self.labels != label).flatten()
+            num_removed = len(self.labels) - len(idx)
```
Comment (jedrzejkozal) [inline utils/buffer.py]: the name of the argument metaclass is misleading.
if metaclass is true you return original labels (fine-grained), otherwise you return meta-labels (metaclass). 
Please use other name for that variable,

File: utils/training.py [hunk 5]
```diff
@@ -141,24 +138,26 @@ def train(model: ContinualModel, dataset: ContinualDataset, args: Namespace) ->
                 if args.debug_mode and i > 3:
                     break
                 if hasattr(dataset.train_loader.dataset, 'logits'):
```
Comment (jedrzejkozal) [inline utils/training.py:137]: Not releted directly to this commit: there is a lot of if else statements in this code right now. In the future we should think about doing something about this. I don't know what this logits attribute of datasets is used for. As far as know Dark Exprience Replay uses the logits that are stored in the buffer. Do you know what this logits in the dataset are for?

**Review by jedrzejkozal — CHANGES_REQUESTED** [review]

File: utils/training.py (hunk 5 above)
Comment (AlifAshrafee) [reply utils/training.py:137]: The logits are used as an additional term in the loss function. According to my understanding, the logits are used as soft targets instead of the actual hard labels. The difference between the logits from the buffer and the current model output is used as the additional loss term.

**Review by AlifAshrafee — COMMENTED** [review]
All issues have been addressed in new commit

**Review by jedrzejkozal — APPROVED** [review]
